crates/ruma-state-res/src/lib.rs

```diff
@@ -167,24 +166,35 @@ fn separate<'a, Id>(
     state_sets_iter: impl Iterator<Item = &'a StateMap<Id>> + Clone,
 ) -> (StateMap<Id>, StateMap<Vec<Id>>)
 where
-    Id: Clone + Eq + 'a,
+    Id: Clone + Eq + Hash + 'a,
 {
+    let mut len: usize = 0;
+    let mut occurrences = HashMap::<_, HashMap<_, usize>>::new();
+
+    for (k, v) in state_sets_iter.clone().flat_map(|x| {
+        len += 1;
+        x
+    }) {
+        occurrences.entry(k).or_default().entry(v).and_modify(|x| *x += 1).or_insert(1);
+    }
+
     let mut unconflicted_state = StateMap::new();
     let mut conflicted_state = StateMap::new();
 
-    for key in state_sets_iter.clone().flat_map(|map| map.keys()).unique() {
-        let mut event_ids =
-            state_sets_iter.clone().map(|state_set| state_set.get(key)).collect::<Vec<_>>();
+    for (k, v) in state_sets_iter.flatten() {
+        let occurrences = occurrences
+            .get(k)
+            .expect("occurrences should contain all state event type and key pairs")
+            .get(v)
+            .expect("occurrences should contain all event IDs");
```

Comment:
```suggestion
    for (k, vs) in occurrences {
        for (v, occurrence_count) in vs {
```

This allows you to avoid cloning the iterator on L174 and avoid cloning the values on L192 and L196. You still have to clone the key, but it's much smaller.

Possibly could swap `occurrence_count` for `occurrences` like the original code, but I find the shadowing a bit confusing here. Similarly, could swap the nested loop for a `flat_map` if you prefer.

Reply:
I changed some variable names like you suggest for clarity. The suggestion to iterate over `occurrences` directly is smart; that saves us from all the `expect`s. Most of the cloning has to stay though, unfortunately, because the function takes borrowed values and returns owned ones.